### DISCUSSION TITLE
Example of IModelConnection.createQueryReader()

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#5579] [Q&A]
I am using ```IModelConnection.query()``` at several locations in my code and after recently upgrading to 3.7.6, it seems this function has been deprecated and replaced by ```createQureyReader()```.   I looked in the iTwin Sample for an example of how to query using this new function but it seems sample there are still using the deprecated function. 

It will be very helpful (and time saver) if I can get a sample of how to use this new function.

Thanks,

Ovais

**Comment by mattbjordan** [comment]
The code examples have been updated in our repository, but not yet updated on [itwinjs.org](https://www.itwinjs.org/). They will be published to the website with the next version release. 

For now, you can find the documentation in our repository at:
https://github.com/iTwin/itwinjs-core/blob/master/docs/learning/ECSQLCodeExamples.md
The code snippets are referenced from another file which can be found at:
https://github.com/iTwin/itwinjs-core/blob/master/example-code/snippets/src/frontend/ExecutingECSQL.ts

You can also check out our 4.0.0 change notes for a brief explanation of how to migrate to `.createQueryReader:`
https://www.itwinjs.org/changehistory/4.0.0/#querying-ecsql

It won't look pretty... but as another option, here is an attached file that is a saved version of the updated web page. This has the code snippets alongside the explanations. You have to download it, change the file extension from `.txt` to `.html` then open it with in a browser.
[ECSQL in iTwin.js Code Examples - Bentley Docs.txt](https://github.com/iTwin/itwinjs-core/files/11615455/ECSQL.in.iTwin.js.Code.Examples.-.Bentley.Docs.txt)

**Reply by mattbjordan** [reply]
I realized that you mentioned upgrading to 3.7.6. For 4.0.0, we made it easier to migrate to `createQueryReader` where you can just replace the `.query` with `.createQueryReader` since the returned ECSqlReader is also an AsyncIterableIterator. But for 3.7.6, that change isn't in there yet.

If possible, I'd recommend upgrading to 4.0.0 as the migration was made much simpler, however...

Here is the changelog (on GitHub) for 3.7.0 which explains how to migrate to `.createQueryReader` for 3.7.x: https://github.com/iTwin/itwinjs-core/blob/release/3.7.0/docs/changehistory/3.7.0.md#itwincore-frontend

`createQueryReader` returns an ECSqlReader object which has a `step` method to iterate over the rows and a `current` property to access the current row during iteration:

So if you had `.query` code that looked like this:
```ts
for await (const row of iModel.query("SELECT * FROM bis.Element", undefined, { rowFormat: QueryRowFormat.UseJsPropertyNames })) {
  const id = row.id;
  // ...
}
```

It now will look like:
```ts
const reader = iModel.createQueryReader("SELECT * FROM bis.Element", undefined, { rowFormat: QueryRowFormat.UseJsPropertyNames });
while (await reader.step()) {
    const id = reader.current.id;
    // ...
}
